### PR TITLE
Improve DNS error handling

### PR DESCRIPTION
**Pull request by dylanratcliffe**

### CODE REVIEW

**Comment by DavidS-ovm** [comment]
ah, ha. Found the rate limiting specs:

> There is a 1024 packet per second (PPS) limit to services that use [link-local](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-instance-addressing.html#link-local-addresses) addresses. This limit includes the aggregate of Route 53 Resolver DNS queries, [Instance Metadata Service (IMDS)](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html) requests, [Amazon Time Service Network Time Protocol (NTP)](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/set-time.html) requests, and [Windows Licensing Service (for Microsoft Windows based instances)](https://aws.amazon.com/windows/resources/licensing/) requests. This quota cannot be increased.

(via https://docs.aws.amazon.com/vpc/latest/userguide/AmazonDNS-concepts.html#vpc-dns-limits)

File: cmd/root.go [hunk 1]
```diff
@@ -72,11 +73,35 @@ var rootCmd = &cobra.Command{
 		healthCheckPort := viper.GetString("service-port")
 		healthCheckPath := "/healthz"
 
+		healthCheckDNSAdapter := adapters.DNSAdapter{}
+
+		// Set up the health check
+		healthCheck := func() error {
+			if !e.IsNATSConnected() {
+				return fmt.Errorf("NATS not connected")
```
[the remark this answers is not in the document]
Comment (DavidS-ovm) [reply cmd/root.go:81]: this will also require adding to the imports above

**Comment by dylanratcliffe** [comment]
That's definitely really high for our purposes right?

File: adapters/dns.go [hunk 2]
```diff
@@ -221,15 +210,67 @@ func (d *DNSAdapter) Search(ctx context.Context, scope string, query string, ign
...
+
+		ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+		defer cancel()
+
+		server := d.GetServers()[i]
+
+		var err error
+		items, err = queryFn(ctx, server)
+
+		if err != nil {
```
Comment (DavidS-ovm) [inline adapters/dns.go:236]:
```suggestion
		items, err = queryFn(ctx, server)
		if err != nil {
```

File: adapters/dns.go [hunk 3]
```diff
@@ -221,15 +210,67 @@ func (d *DNSAdapter) Search(ctx context.Context, scope string, query string, ign
...
+				strings.Contains(err.Error(), "temporary failure") {
+				return err // Retry on timeout
+			}
+			return backoff.Permanent(err)
+		}
+
+		return nil
+	}
 
+	err := backoff.Retry(operation, b)
```
Comment (DavidS-ovm) [inline adapters/dns.go:250]:
```suggestion
	err := backoff.Retry(operation, b)
  span := trace.SpanFromContext(ctx)
  span.SetAttributes(
    attributes.String("ovm.dns.server", server)
  )
```

**Comment by DavidS-ovm** [comment]
> That's definitely really high for our purposes right?

Yeah, we're far away from that limit, but it is also an order of magnitude less than I thought it was. And hitting it will impact everything in our VPC, so it's something to keep in mind for the future.

**Review by DavidS-ovm — APPROVED** [review]

**Pull request merged**